src/Compilers/Core/Portable/InternalUtilities/FailFast.cs

```diff
@@ -32,12 +32,39 @@ internal static void OnFatalException(Exception exception)
             {
                 exception = aggregate.InnerExceptions[0];
             }
+
 #endif
+            DumpStackTrace(exception);
 
             Environment.FailFast(exception.ToString(), exception);
             throw ExceptionUtilities.Unreachable; // to satisfy [DoesNotReturn]
         }
 
+        /// <summary>
+        /// Dumps the stack trace of the exception and the handler to the console. This is useful
+        /// for debugging unit tests that hit a fatal exception
+        /// </summary>
+        [Conditional("DEBUG")]
+        private static void DumpStackTrace(Exception exception)
+        {
+            Console.WriteLine("Dumping info before call to failfast");
+            Console.WriteLine("Exception info");
+            Exception? current = exception;
+            do
+            {
+                Console.WriteLine(current.Message);
+                Console.WriteLine(current.StackTrace);
+                current = current.InnerException;
+            } while (current is object);
```

Comment:
a normal `while` loops seems like it would work just as well (and be clearer).  Alternatively, you can do:

```
for (var current = exception; current != null; current = current.InnerException) {
    ...
}
```

As this is the standard 'walk a linked list' idiom.